primitives/timestamp/src/lib.rs

```diff
@@ -240,34 +239,8 @@ impl sp_inherents::InherentDataProvider for InherentDataProvider {
 		identifier: &InherentIdentifier,
 		error: &[u8],
 	) -> Option<Result<(), sp_inherents::Error>> {
-		if *identifier != INHERENT_IDENTIFIER {
-			return None
-		}
-
-		match InherentError::try_from(&INHERENT_IDENTIFIER, error)? {
-			InherentError::ValidAtTimestamp(valid) => {
-				let max_drift = self.max_drift;
-				let timestamp = self.timestamp;
-				// halt import until timestamp is valid.
-				// reject when too far ahead.
-				if valid > timestamp + max_drift {
-					return Some(Err(sp_inherents::Error::Application(Box::from(
-						InherentError::TooFarInFuture,
-					))))
-				}
-
-				let diff = valid.checked_sub(timestamp).unwrap_or_default();
-				log::info!(
-					target: "timestamp",
-					"halting for block {} milliseconds in the future",
-					diff.0,
-				);
-
-				futures_timer::Delay::new(diff.as_duration()).await;
-
-				Some(Ok(()))
-			},
-			o => Some(Err(sp_inherents::Error::Application(Box::from(o)))),
-		}
+		Some(Err(sp_inherents::Error::Application(Box::from(InherentError::try_from(
```

Comment:
If I understood correctly the `InherentDataProvider` for tuples [here](https://github.com/paritytech/substrate/blob/2ced413e40eed0e294fbc4c0d539bac2376016f6/primitives/inherents/src/client_side.rs#L121-L123) stops trying handling the error as soon as one of the tuple elements returns `Some`. 

That is why every implementation we have immediately returns `None` if `identifier != INHERENT_DATA_IDENTIFIER`.

Maybe we should we leave the check here as well? Otherwise we will return `Some` even if the "notification" is addressed to the Timestamp data provider

Reply:
The `try_from` implementation of the error is doing this check internally. :)

Reply:
Got it. Thank you

Reply:
But isn't the result `Some` anyway? The `None` would be stored within the `Application` variant right?

Reply:
Ignore my commend (that I've cancelled ) I've expanded the macro 😃 and now is clear Ty